### PR TITLE
Image preview background contrast

### LINKED ISSUE #319: Detect contrast colour for downloaded images
## :warning: Is your feature request related to a problem? Please describe

Downloading images that are fully tinted with a colour in a range of the background colour used by Chucker makes them virtually invisible for the user. Here is an example of such an icon which is white. It is visible in night mode but not in day mode.

## :bulb: Describe the solution you'd like

Use `Palette` to determine what should be a contrasting background for the image. It could look like something along the lines of this.

https://github.com/MiSikora/Squash-It/blob/93812d1a13d8fbb62c2290659d0785f250f94507/squash-it/library/src/main/java/io/mehow/squashit/screenshot/ScreenshotActivity.kt#L55-L60

The only difference would be to use `Palette` output to get colour instead of day or night mode.

## :page_facing_up: Additional context

Screenshot_20200407-172130

Screenshot_20200407-172209

## :raising_hand: Do you want to develop this feature yourself?

- [x] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by MiSikora**
## :camera: Screenshots

Screenshot_1586783170
Screenshot_1586783151
Screenshot_1586783159
Screenshot_1586783154
Screenshot_1586783146
Screenshot_1586783173

## :page_facing_up: Context

There is an issue. #319

## :pencil: Changes

Showing downloaded image preview triggers `Palette` to determine contrast colour based on the luminescence of the image. You can see in the screenshots that light image has dark background and vice versa.

One unfortunate thing is that `Palette` does not recognise alpha channel so we cannot filter out the the transparent background and `Palette` treats it as a black colour. What I do is I create a new bitmap that uses magenta in place of transparent pixels and then filter them out in the `Palette`. I chose magenta as it is rather a rare colour.

## :no_entry_sign: Breaking

No.

## :hammer_and_wrench: How to test

I added image requests to the sample app. Unfortunately I didn't find any image generator that would operate on images with alpha and starting a local server just for that purpose seems too much.

## :stopwatch: Next steps

Closes #319.

### CODE REVIEW

**Comment by vbuberen** [comment]
I haven't worked with Palette before, but is it that demanding task that it requires something to offload it from the main thread?

**Comment by MiSikora** [comment]
> 1. I don't like that we have that color defining logic inside adapter. While added code relates to displaying I would try to avoid doing such things there and leave only things which adapter usually does there (I am also struggling with such issue now with my payload search improvements feature)
>    Since we already doing our own items creation for this adapter in PayloadFragment, let's do this thing there as well and just pass the background color into `TransactionPayloadItem`. We already have a coroutine there and you shouldn't worry about creating new async things.

Sure, I'll change it.

> 2. Users might not be aware whenever this contrast border is a part of their downloaded image or Chucker's background. This is solely a UX problem, but quite serious from my point of view.  I will think about ideas here, but we need to clearly indicate our background and user content to not confuse anyone.

I was thinking about changing the whole theme for a fragment but it would make a little bit inconsistent with other ones. If it is not an issue for you I'd go with that.

> I haven't worked with Palette before, but is it that demanding task that it requires something to offload it from the main thread?

It iterates over all pixels of an image and analyses them in a histogram-like fashion. It depends on the size of the image and the max area of the image that you allow for. By default it limits it to `112 x 112`. Personally, I wouldn't bother with doing it asynchronously but didn't know what's your approach so I took safer one.

**Comment by vbuberen** [comment]
As to background - you can leave it for now. I might come up with some ideas here, so to not make you do something twice let's skip for now.

Thanks for explanations about Palette stuff.

**Comment by MiSikora** [comment]
> Thanks for explanations about Palette stuff.

Just do be sure. You'd like to change it to synchronous call? I can always make it asynchronous and write a suspending extension since it will be moved out from the `ViewHolder` where there is access to some UI scope.

**Comment by vbuberen** [comment]
No-no, under `background` I meant fragment background change that your mentioned when told about theme 😄

**Comment by vbuberen** [comment]
That code with palette can would perfectly fit the coroutine in `processPayload`, so let it be there.

**Comment by cortinico** [comment]
> Users might not be aware whenever this contrast border is a part of their downloaded image or Chucker's background. This is solely a UX problem, but quite serious from my point of view. I will think about ideas here, but we need to clearly indicate our background and user content to not confuse anyone.

I'm also a bit concerned here. How about having an icon on the bottom right (or somewhere else) that will toggle the background of the `ImageView` from black to white and vice versa?

I really like the approach of using `Palette` but I'm afraid it could create some confusion in the user that don't really understand what is going on (plus is yet another dependency that we're adding on our users load).

**Comment by MiSikora** [comment]
> I'm also a bit concerned here. How about having an icon on the bottom right (or somewhere else) that will toggle the background of the `ImageView` from black to white and vice versa?

The only issue is that it requires one step from the user which might not be at time confusing is when they do not see the image. Maybe a switch (with i.e. moon and sun icons) that triggers background of the image view and that has the initial state set by `Palette`?

> I really like the approach of using `Palette` but I'm afraid it could create some confusion in the user that don't really understand what is going on (plus is yet another dependency that we're adding on our users load).

 I wouldn't be too concerned about the size of `Palette`. It has like 160 methods and 23 KB. Unless you also mean additional dependency that devs have to be aware of. But then, I guess, automatic recognition is out of the scope and we should go only with the switch.

**Comment by cortinico** [comment]
> Unless you also mean additional dependency that devs have to be aware of.

I'm just afraid that pulling in dependencies could create problems in debug builds for our consumers. Say a customer is on a older version of `Palette`, and applying `Chucker` that it's on a newer version of `Palette` will cause the resulting version to be ours. This could have unintended side effects as well as complicating debugging.

Anyway I feel that I'm maybe worrying too much for this specific case as the library is pretty small.

> I wouldn't be too concerned about the size of `Palette`. It has like 160 methods and 23 KB. 

Ok it feels like is something we can live with. Let's go for that.

> Maybe a switch (with i.e. moon and sun icons) that triggers background of the image view and that has the initial state set by Palette?

This would be my preferred approach if possible 👍

**Comment by vbuberen** [comment]
@cortinico 
Regarding concerns about versions difference between Chucker and user's app - I am not sure a lot of people use `Palette` in their dependencies. Also, it seems to update not that often, so we should be fine here.
As to additional button - I am not sure where we should place it at the moment.

What comes to my mind is a variation of idea suggested initially by @MiSikora on background change, but only for image view item - to make such item match parent view width and height and make it is background grey (for example) so it will be clear that it is our background. 
Here is a quick prototype of such thing:
Screenshot 2020-04-13 at 14 22 34
It is not ideal, but might be better for user perception. We might need to increase height as well.

What do you think, guys?

**Comment by vbuberen** [comment]
Or we might take it further and in that `Response` tab add some image data, like its resolution, etc. and after a few sort of headers add something like `Image preview:` item and place an image below. Might be a good enhancement. 
It is not a request for this PR or author of this PR 😄Just came to my mind.

**Comment by MiSikora** [comment]
> As to additional button - I am not sure where we should place it at the moment.

Yes, I'm also having trouble picturing it. I tried to code it but then I kind of hit a brick wall when it came to the design. Mostly because images can take up the whole width of the screen. I thought of two solutions.

1. There is a button below the image that changes the background of the whole item.
2. There is a switch to the end of an item. But then item should be probably a constraint layout that limits the size of an image to i.e. 60% of width so it still can be centred and fit the switch at the same time.

> What comes to my mind is a variation of idea suggested initially by @MiSikora on background change, but only for image view item - to make such item match parent view width and height and make it is background grey (for example) so it will be clear that it is our background.
> 
> What do you think, guys?

Do you mean to have it always grey regardless of the luminescence? I think it might be ok but maybe it would be better just to have two values of grey? One that is good for light images and the other that is good for dark images. Otherwise it will fall into the same trap as the original issue if an image has a transparent background and is tinted with a colour similar to the background one.

> Or we might take it further and in that `Response` tab add some image data, like its resolution, etc. and after a few sort of headers add something like `Image preview:` item and place an image below. Might be a good enhancement.

It would be good for now but I think this should belong to an item in a view holder if multipart support is something to be implemented in the future.

**Comment by vbuberen** [comment]
> Mostly because images can take up the whole width of the screen.

I would like to address it in some future PRs by adding sort of scaling for images.

> Do you mean to have it always grey regardless of the luminescence? I think it might be ok but maybe it would be better just to have two values of grey?

No, we would have to use 2 shades of grey anyway to match both day and night theme or image luminescence. 
And another idea just came to me after you mentioned images with alpha channel - what if we use patterned background for images, like Android Studio does (attaching a screenshot)? This background would be hard to confuse with image content, I guess.

Screenshot 2020-04-13 at 14 57 49

**Comment by MiSikora** [comment]
> And another idea just came to me after you mentioned images with alpha channel - what if we use patterned background for images, like Android Studio does (attaching a screenshot)? This background would be hard to mix with image content, I guess.

Yes, I like. I will make another PR with this approach to see how it looks like.

**Comment by MiSikora** [comment]
I added the chessboard background. I think the result is good. The colours for chess fields are based on the primary colour so they are not only in shades of grey. Here are the screenshots.

Screenshot_1586783170
Screenshot_1586783151
Screenshot_1586783159
Screenshot_1586783154
Screenshot_1586783146
Screenshot_1586783173

File: library/src/main/java/com/chuckerteam/chucker/internal/support/ChessboardDrawable.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package com.chuckerteam.chucker.internal.support
```
Comment (MiSikora) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/ChessboardDrawable.kt:1]: I wasn't sure about the package for this class. I added it to the `support` but maybe `ui` is more appropriate.

Comment (vbuberen) [reply library/src/main/java/com/chuckerteam/chucker/internal/support/ChessboardDrawable.kt:1]: From my point of view it is fine to place in `support`.

**Comment by cortinico** [comment]
> The result with that chessboard pattern is awesome 👍

Really great job 💪 Thanks for the addition @MiSikora

**Pull request merged**